src/App.tsx

```diff
@@ -1,24 +1,234 @@
+      userId: USER_ID,
+      completed: false,
+    };
+
+    try {
+      setIsTitleDisabled(true);
+
+      const newTodo = await addTodo(USER_ID, todoToAdd);
+
+      setIsProcessedIds(currIds => [...currIds, newTodo.id]);
+      setTempTodo(newTodo);
+
+      const loadedTodos = await getTodos(USER_ID);
+
+      setTodos(loadedTodos);
+    } catch (error) {
+      setHasError(true);
+      setErrorType(ErrorType.Add);
+      clearErrorAfterDelay();
+    } finally {
+      setTitle('');
+      setTempTodo(null);
+      setIsTitleDisabled(false);
+      setIsProcessedIds([]);
+    }
+  }, []);
+
+  const handleDeleteTodo = useCallback(async (todoId: number) => {
+    try {
+      setIsProcessedIds(currIds => [...currIds, todoId]);
+      await removeTodo(USER_ID, todoId);
+
+      setTodos(currTodos => currTodos.filter(todo => todo.id !== todoId));
+    } catch (error) {
+      setHasError(true);
+      setErrorType(ErrorType.Delete);
+      clearErrorAfterDelay();
+    } finally {
+      setIsProcessedIds([]);
+    }
+  }, []);
+
+  const handleCompletedTodos = useCallback(async () => {
+    try {
+      await Promise.all(todos.map(async todo => {
+        if (todo.completed) {
+          await handleDeleteTodo(todo.id);
+        }
+      }));
+    } catch (error) {
+      setErrorType(ErrorType.Update);
+      setHasError(true);
+    }
+  }, [todos]);
+
+  const handleUpdateTodo = useCallback(
+    async (todoId: number, value: boolean | string) => {
+      try {
+        let newData = {};
+
+        if (typeof value === 'boolean') {
+          newData = {
+            completed: value,
+          };
+        }
+
+        if (typeof value === 'string') {
+          newData = {
+            title: value,
+          };
+        }
+
+        setIsProcessedIds(currIds => [...currIds, todoId]);
+
+        await updateTodo(USER_ID, todoId, newData);
+        await fetchAllTodos();
+      } catch (error) {
+        setHasError(true);
+        setErrorType(ErrorType.Update);
+        clearErrorAfterDelay();
+      } finally {
+        setIsProcessedIds([]);
+      }
+    }, [],
+  );
+
+  const handleAllTodosStatus = async () => {
+    try {
+      const areTodosCompleted = todos.every(todo => todo.completed);
+
+      todos.forEach(async todo => {
+        await handleUpdateTodo(todo.id, !areTodosCompleted);
+      });
+    } catch (error) {
+      setErrorType(ErrorType.Update);
+      setHasError(true);
+    }
+  };
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+
+        <Header
```

Comment:
add Loader here while we waiting for todos fetched

Reply:
I'm not sure if this is correct, but in my case the Loader shows up every time when updating a todo (changing title or status)

Reply:
@Coyoote that's because you should return all components when loading false:
```
  {isLoading
         ? <Loader />
         : (
           <div className="todoapp__content">
             <Header />

             <TodoList />

             {!!todos.length && (
               <Footer />
             )}
           </div>
         )}
```